oeps/oep-0003.rst

```diff
@@ -0,0 +1,144 @@
++---------------+-------------------------------------------+
+| Author        | Jeremy Bowman <jbowman@edx.org>           |
++---------------+-------------------------------------------+
+| Arbiter       |                                           |
++---------------+-------------------------------------------+
+| Status        | Draft                                     |
++---------------+-------------------------------------------+
+| Type          | Architecture                              |
++---------------+-------------------------------------------+
+| Created       | 2016-06-09                                |
++---------------+-------------------------------------------+
+
+Abstract
+========
+
+Proposes a system which allows users to initiate, monitor, and retrieve the
+results of various long-running tasks in a consistent manner.  This system
+leverages existing libraries where practical and makes the development of new
+asynchronous tasks straightforward.
+
+
+Specification
+=============
+
+A new open source Django application will be written to support a user
+interface for managing celery tasks in a consistent manner across services.
+The main goals of the application are, roughly in order:
+
+1. Allow a long-running task triggered by a user to be run asynchronously,
+   while it is clear to the user that they are free to do other work on the
+   site in the meantime
+2. Allow the user to fetch any output of the task after it finishes
+3. Notify the user when the task is complete
+4. Give the user some indication of how far along an in-progress task is
+5. Support prevention of multiple identical tasks from executing concurrently
+
+Celery will be used to run the asynchronous tasks, but the design should be
+capable of adapting to a different task runner later if necessary.  Database
+models will be used to represent tasks as well as their input, output, and
+current status.  It will be possible to present a fan-out of Celery tasks
+instantiated to perform work in parallel as a single task from the user's
+perspective with a meaningful progress indicator.
+
+For notifications, email is sufficient for the initial implementation (but
+other push notifications may be added later).  A task status page for each
+task will show any available information about how far along the task is, and
+allow retrieval of any output it may have generated.  The user will be
+redirected to this status page when the task is requested.  The initial
+implementation might not include real-time updates of the data on the page or
+a dashboard listing all tasks initiated by the user, but these will be added
+in time.  Similarly, support for tasks which can be monitored by multiple
+users with sufficient access priveleges is desired but may be added after the
+initial release.
+
+Course import and export will utilize the new system when it is released, and
+other long-running tasks will switch to use it as needed.
+
+Motivation
+==========
+
```

Comment:
Great question for @cpennington, who created this template with sections in the order you're seeing here. I had a similar question while thinking through my own OEP, and agree that some of the underlying "why do you care about this" stuff should bubble up closer to the abstract.